File: app/src/main/java/vn/loitp/app/activity/function/viewdraghelpersimple/VDHView.java
package vn.loitp.app.activity.function.viewdraghelpersimple;

import android.content.Context;
import android.support.annotation.AttrRes;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.view.MotionEventCompat;
import android.support.v4.view.ViewCompat;
import android.support.v4.widget.ViewDragHelper;
import android.util.AttributeSet;
import android.util.Log;
import android.view.MotionEvent;
import android.view.View;
import android.widget.LinearLayout;

import loitp.basemaster.R;
import vn.loitp.core.utilities.LLog;
import vn.loitp.core.utilities.LScreenUtil;

public class VDHView extends LinearLayout {
    private final String TAG = getClass().getSimpleName();
    private View headerView;
    private View bodyView;
    private ViewDragHelper mViewDragHelper;
    private int mAutoBackViewX;
    private int mAutoBackViewY;
    //private float mInitialMotionX;
    //private float mInitialMotionY;
    private int mDragRange;
    private float mDragOffset;
    private boolean isEnableAlpha = true;
    private boolean isEnableRevertMaxSize = true;
    private boolean isMinimized;//header view is scaled at least 1
    private int sizeWHeaderViewOriginal;
    private int sizeHHeaderViewOriginal;
    private int sizeWHeaderViewMin;
    private int sizeHHeaderViewMin;
    private int newSizeWHeaderView;
    private int newSizeHHeaderView;
    private int mCenterY;
    private int mCenterX;
    private int screenW;
    private int screenH;

    public VDHView(@NonNull Context context) {
        this(context, null);
    }

    public VDHView(@NonNull Context context, @Nullable AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public VDHView(@NonNull Context context, @Nullable AttributeSet attrs, @AttrRes int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        initView();
    }

    public interface Callback {
        public void onStateChange(State state);

        public void onPartChange(Part part);

        public void onViewPositionChanged(int left, int top, float dragOffset);

        public void onOverScroll(State state, Part part);
    }

    private Callback callback;

    public void setCallback(Callback callback) {
        this.callback = callback;
    }

    private void initView() {
        mViewDragHelper = ViewDragHelper.create(this, 1.0f, mCallback);
        mViewDragHelper.setEdgeTrackingEnabled(ViewDragHelper.EDGE_LEFT);
    }

    @Override
    protected void onFinishInflate() {
        super.onFinishInflate();
        screenW = LScreenUtil.getScreenWidth();
        screenH = LScreenUtil.getScreenHeight();
        headerView = findViewById(R.id.header_view);
        bodyView = findViewById(R.id.body_view);
        headerView.post(new Runnable() {
            @Override
            public void run() {
                sizeWHeaderViewOriginal = headerView.getMeasuredWidth();
                sizeHHeaderViewOriginal = headerView.getMeasuredHeight();
                sizeWHeaderViewMin = sizeWHeaderViewOriginal / 2;
                sizeHHeaderViewMin = sizeHHeaderViewOriginal / 2;
                //LLog.d(TAG, "size original: " + sizeWHeaderViewOriginal + "x" + sizeHHeaderViewOriginal + " -> " + sizeWHeaderViewMin + "x" + sizeHHeaderViewMin);
            }
        });
    }

    private ViewDragHelper.Callback mCallback = new ViewDragHelper.Callback() {
        @Override
        public void onViewPositionChanged(@NonNull View changedView, int left, int top, int dx, int dy) {
            super.onViewPositionChanged(changedView, left, top, dx, dy);
            if (mDragOffset == (float) top / mDragRange) {
                return;
            } else {
                mDragOffset = (float) top / mDragRange;
            }
            if (mDragOffset > 1) {
                mDragOffset = 1;
            }
            if (mDragOffset < 0) {
                mDragOffset = 0;
            }
            if (callback != null) {
                callback.onViewPositionChanged(left, top, mDragOffset);
            }

            int x = 0;
            int y = headerView.getHeight() + top;
            bodyView.layout(x, y, x + bodyView.getMeasuredWidth(), y + bodyView.getMeasuredHeight());
            if (isEnableAlpha) {
                bodyView.setAlpha(1 - mDragOffset / 2);
            } else {
                if (bodyView.getAlpha() != 1f) {
                    bodyView.setAlpha(1f);
                }
            }

            if (isMinimized) {
                if (isEnableRevertMaxSize) {
                    headerView.setPivotX(headerView.getWidth() / 2f);
                    headerView.setPivotY(headerView.getHeight());
                    headerView.setScaleX(1 - mDragOffset / 2);
                    headerView.setScaleY(1 - mDragOffset / 2);
                } else {

                }
            } else {
                headerView.setPivotX(headerView.getWidth() / 2f);
                headerView.setPivotY(headerView.getHeight());
                headerView.setScaleX(1 - mDragOffset / 2);
                headerView.setScaleY(1 - mDragOffset / 2);
            }

            newSizeWHeaderView = (int) (sizeWHeaderViewOriginal * headerView.getScaleX());
            newSizeHHeaderView = (int) (sizeHHeaderViewOriginal * headerView.getScaleY());

            mCenterX = left + sizeWHeaderViewOriginal / 2;
            mCenterY = top + newSizeHHeaderView / 2 + sizeHHeaderViewOriginal - newSizeHHeaderView;

            //int posX = centerX - newSizeWHeaderView / 2;
            //int posY = centerY - newSizeHHeaderView / 2;
            //LLog.d(TAG, "onViewPositionChanged left: " + left + ", top: " + top + ", mDragOffset: " + mDragOffset + " => newSizeW " + newSizeWHeaderView + "x" + newSizeHHeaderView + "=> centerX: " + centerX + ", centerY: " + centerY + ", posX: " + posX + ", posY: " + posY);
            LLog.d(TAG, "onViewPositionChanged left: " + left + ", top: " + top + ", mDragOffset: " + mDragOffset + " => newSizeW " + newSizeWHeaderView + "x" + newSizeHHeaderView + "=> mCenterX: " + mCenterX + ", mCenterY: " + mCenterY);

            if (mDragOffset == 0) {
                //top_left, top, top_right
                if (left <= -headerView.getWidth() / 2) {
                    changeState(State.TOP_LEFT);
                } else if (left >= headerView.getWidth() / 2) {
                    changeState(State.TOP_RIGHT);
                } else {
                    changeState(State.TOP);
                }
            } else if (mDragOffset == 1) {
                //bottom_left, bottom, bottom_right
                if (left <= -headerView.getWidth() / 2) {
                    changeState(State.BOTTOM_LEFT);
                } else if (left >= headerView.getWidth() / 2) {
                    changeState(State.BOTTOM_RIGHT);
                } else {
                    changeState(State.BOTTOM);
                }
                isMinimized = true;
            } else {
                //mid_left, mid, mid_right
                if (left <= -headerView.getWidth() / 2) {
                    changeState(State.MID_LEFT);
                } else if (left >= headerView.getWidth() / 2) {
                    changeState(State.MID_RIGHT);
                } else {
                    changeState(State.MID);
                }
            }
            if (mCenterY < screenH / 2) {
                if (mCenterX < screenW / 2) {
                    changePart(Part.TOP_LEFT);
                } else {
                    changePart(Part.TOP_RIGHT);
                }
            } else {
                if (mCenterX < screenW / 2) {
                    changePart(Part.BOTTOM_LEFT);
                } else {
                    changePart(Part.BOTTOM_RIGHT);
                }
            }
        }

        @Override
        public boolean tryCaptureView(View child, int pointerId) {
            return headerView == child;
        }

        @Override
        public int clampViewPositionVertical(@NonNull View child, int top, int dy) {
            int minY;
            if (isEnableRevertMaxSize) {
                minY = -child.getHeight() / 2;
            } else {
                minY = -sizeHHeaderViewMin * 3 / 2;
            }
            float scaledY = child.getScaleY();
            int sizeHScaled = (int) (scaledY * child.getHeight());
            int maxY = getHeight() - sizeHScaled * 3 / 2;
            //LLog.d(TAG, "clampViewPositionVertical getHeight:" + getHeight() + ", scaledY:" + scaledY + ", sizeHScaled: " + sizeHScaled + ", top:" + top + ", minY: " + minY + ", maxY: " + maxY);
            if (top <= minY) {
                return minY;
            } else if (top > maxY) {
                return maxY;
            } else {
                return top;
            }
        }

        @Override
        public int clampViewPositionHorizontal(@NonNull View child, int left, int dx) {
            int minX = -child.getWidth() / 2;
            int maxX = getWidth() - child.getWidth() / 2;
            if (left <= minX) {
                return minX;
            } else if (left > maxX) {
                return maxX;
            } else {
                return left;
            }
        }

        @Override
        public void onViewReleased(View releasedChild, float xvel, float yvel) {
            if (releasedChild == headerView) {
                mViewDragHelper.settleCapturedViewAt(mAutoBackViewX, mAutoBackViewY);
            }
            invalidate();
        }

        @Override
        public void onEdgeDragStarted(int edgeFlags, int pointerId) {
        }
    };

    private void changeState(State newState) {
        if (state != newState) {
            state = newState;
            LLog.d(TAG, "changeState: " + newState);
            if (callback != null) {
                callback.onStateChange(state);
            }
        }
    }

    private void changePart(Part newPart) {
        if (part != newPart) {
            part = newPart;
            LLog.d(TAG, "changePart: " + part);
            if (callback != null) {
                callback.onPartChange(part);
            }
        }
    }

    @Override
    public void computeScroll() {
        if (mViewDragHelper.continueSettling(true)) {
            ViewCompat.postInvalidateOnAnimation(this);
        }
    }

    @Override
    public boolean onInterceptTouchEvent(MotionEvent ev) {
        return mViewDragHelper.shouldInterceptTouchEvent(ev);
    }

    @Override
    public boolean onTouchEvent(MotionEvent event) {
        mViewDragHelper.processTouchEvent(event);
        final float x = event.getX();
        final float y = event.getY();
        boolean isViewUnder = mViewDragHelper.isViewUnder(headerView, (int) x, (int) y);
        //LLog.d(TAG, "onTouchEvent isViewUnder: " + isViewUnder);
        switch (event.getAction() & MotionEventCompat.ACTION_MASK) {
            /*case MotionEvent.ACTION_DOWN: {
                LLog.d(TAG, "onTouchEvent ACTION_DOWN touch:" + x + " x " + y);
                LLog.d(TAG, "onTouchEvent ACTION_DOWN isViewUnder:" + isViewUnder);
                break;
            }*/
            case MotionEvent.ACTION_UP: {
                LLog.d(TAG, "onTouchEvent ACTION_UP state:" + state.name() + ", mCenterX: " + mCenterX);
                if (state == State.TOP_LEFT || state == State.TOP_RIGHT || state == State.BOTTOM_LEFT || state == State.BOTTOM_RIGHT) {
                    LLog.d(TAG, "destroy state: " + state.name());
                    if (callback != null) {
                        callback.onOverScroll(state, part);
                    }
                } else {
                    if (part == Part.BOTTOM_LEFT) {
                        minimizeBottomLeft();
                    } else if (part == Part.BOTTOM_RIGHT) {
                        minimizeBottomRight();
                    } else if (part == Part.TOP_LEFT) {
                        if (isEnableRevertMaxSize) {
                            maximize();
                        } else {
                            if (isMinimized) {
                                minimizeTopLeft();
                            }
                        }
                    } else if (part == Part.TOP_RIGHT) {
                        if (isEnableRevertMaxSize) {
                            maximize();
                        } else {
                            if (isMinimized) {
                                minimizeTopRight();
                            }
                        }
                    }
                }
                break;
            }
        }
        return isViewUnder;
    }

    @Override
    protected void onLayout(boolean changed, int l, int t, int r, int b) {
        super.onLayout(changed, l, t, r, b);
        mDragRange = getHeight() - headerView.getHeight();
        mAutoBackViewX = headerView.getLeft();
        mAutoBackViewY = headerView.getTop();
        //LLog.d(TAG, "onLayout l:" + l + ", t:" + t + ", r:" + r + ", b:" + b + ", mAutoBackViewX: " + mAutoBackViewX + ", mAutoBackViewY: " + mAutoBackViewY);
    }

    public enum State {TOP, TOP_LEFT, TOP_RIGHT, BOTTOM, BOTTOM_LEFT, BOTTOM_RIGHT, MID, MID_LEFT, MID_RIGHT}

    public enum Part {TOP_LEFT, TOP_RIGHT, BOTTOM_LEFT, BOTTOM_RIGHT}

    private State state;
    private Part part;

    public State getState() {
        return state;
    }

    public Part getPart() {
        return part;
    }

    public void maximize() {
        if (isEnableRevertMaxSize) {
            smoothSlideTo(0, 0);
        } else {
            Log.e(TAG, "Error: cannot maximize because isEnableRevertMaxSize is true");
        }
    }

    public void minimizeBottomLeft() {
        int posX = getWidth() - sizeWHeaderViewOriginal - sizeWHeaderViewMin / 2;
        int posY = getHeight() - sizeHHeaderViewOriginal;
        //LLog.d(TAG, "minimize " + posX + "x" + posY);
        smoothSlideTo(posX, posY);
    }

    public void minimizeBottomRight() {
        int posX = getWidth() - sizeWHeaderViewOriginal + sizeWHeaderViewMin / 2;
        int posY = getHeight() - sizeHHeaderViewOriginal;
        //LLog.d(TAG, "minimize " + posX + "x" + posY);
        smoothSlideTo(posX, posY);
    }

    public void minimizeTopRight() {
        if (isEnableRevertMaxSize) {
            Log.e(TAG, "Error: cannot minimizeTopRight because isEnableRevertMaxSize is true");
            return;
        }
        if (!isMinimized) {
            Log.e(TAG, "Error: cannot minimizeTopRight because isMinimized is false. This function only works if the header view is scrolled BOTTOM");
            return;
        }
        int posX = screenW - sizeWHeaderViewMin * 3 / 2;
        int posY = -sizeHHeaderViewMin;
        //LLog.d(TAG, "minimizeTopRight " + posX + "x" + posY);
        smoothSlideTo(posX, posY);
    }

    public void minimizeTopLeft() {
        if (isEnableRevertMaxSize) {
            Log.e(TAG, "Error: cannot minimizeTopRight because isEnableRevertMaxSize is true");
            return;
        }
        if (!isMinimized) {
            Log.e(TAG, "Error: cannot minimizeTopRight because isMinimized is false. This function only works if the header view is scrolled BOTTOM");
            return;
        }
        int posX = -sizeWHeaderViewMin / 2;
        int posY = -sizeHHeaderViewMin;
        //LLog.d(TAG, "minimizeTopRight " + posX + "x" + posY);
        smoothSlideTo(posX, posY);
    }

    public void smoothSlideTo(int positionX, int positionY) {
        if (mViewDragHelper.smoothSlideViewTo(headerView, positionX, positionY)) {
            ViewCompat.postInvalidateOnAnimation(this);
            postInvalidate();
        }
    }

    public boolean isMinimized() {
        return isMinimized;
    }

    public boolean isEnableAlpha() {
        return isEnableAlpha;
    }

    public void setEnableAlpha(boolean enableAlpha) {
        isEnableAlpha = enableAlpha;
    }

    public boolean isEnableRevertMaxSize() {
        return isEnableRevertMaxSize;
    }

    public void setEnableRevertMaxSize(boolean enableRevertMaxSize) {
        isEnableRevertMaxSize = enableRevertMaxSize;
    }

    public void toggleShowHideHeaderView() {
        if (headerView.getVisibility() == VISIBLE) {
            headerView.setVisibility(INVISIBLE);
        } else {
            headerView.setVisibility(VISIBLE);
        }
    }

    public void toggleShowHideBodyView() {
        if (bodyView.getVisibility() == VISIBLE) {
            bodyView.setVisibility(INVISIBLE);
        } else {
            bodyView.setVisibility(VISIBLE);
        }
    }

    public void onPause() {
        if (!isEnableRevertMaxSize) {
            headerView.setVisibility(INVISIBLE);
            bodyView.setVisibility(INVISIBLE);
            minimizeBottomRight();
        }
    }

    public void dissappear() {

    }

    public void appear() {

    }
}

File: app/src/main/java/vn/loitp/up/a/func/viewDragHelperSimple/VDHView.kt
package vn.loitp.up.a.func.viewDragHelperSimple

import android.annotation.SuppressLint
import android.content.Context
import android.util.AttributeSet
import android.view.MotionEvent
import android.view.View
import android.widget.LinearLayout
import androidx.annotation.AttrRes
import androidx.core.view.MotionEventCompat
import androidx.core.view.ViewCompat
import androidx.customview.widget.ViewDragHelper
import com.loitp.core.ext.screenHeight
import com.loitp.core.ext.screenWidth
import vn.loitp.R

class VDHView @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet? = null,
    @AttrRes defStyleAttr: Int = 0
) : LinearLayout(context, attrs, defStyleAttr) {

    private lateinit var headerView: View
    private lateinit var bodyView: View
    private lateinit var mViewDragHelper: ViewDragHelper
    private var mAutoBackViewX = 0
    private var mAutoBackViewY = 0

    // private float mInitialMotionX;
    // private float mInitialMotionY;
    private var mDragRange = 0
    private var mDragOffset = 0f
    var isEnableAlpha = true
    var isEnableRevertMaxSize = true
    var isMinimized = false // header view is scaled at least 1 = false
    private var sizeWHeaderViewOriginal = 0
    private var sizeHHeaderViewOriginal = 0
    private var sizeWHeaderViewMin = 0
    private var sizeHHeaderViewMin = 0
    private var newSizeWHeaderView = 0
    private var newSizeHHeaderView = 0
    private var mCenterY = 0
    private var mCenterX = 0
    private var screenW = 0
    private var screenH = 0

    interface Callback {
        fun onStateChange(state: State)
        fun onPartChange(part: Part)
        fun onViewPositionChanged(left: Int, top: Int, dragOffset: Float)
        fun onOverScroll(state: State?, part: Part?)
    }

    private var callback: Callback? = null

    fun setCallback(callback: Callback?) {
        this.callback = callback
    }

    private fun initView() {
        mViewDragHelper = ViewDragHelper.create(this, 1.0f, mCallback)
        mViewDragHelper.setEdgeTrackingEnabled(ViewDragHelper.EDGE_LEFT)
    }

    override fun onFinishInflate() {
        super.onFinishInflate()

        screenW = screenWidth
        screenH = screenHeight

        headerView = findViewById(R.id.headerView)
        bodyView = findViewById(R.id.bodyView)

        headerView.post {
            sizeWHeaderViewOriginal = headerView.measuredWidth
            sizeHHeaderViewOriginal = headerView.measuredHeight
            sizeWHeaderViewMin = sizeWHeaderViewOriginal / 2
            sizeHHeaderViewMin = sizeHHeaderViewOriginal / 2
        }
    }

    private val mCallback: ViewDragHelper.Callback = object : ViewDragHelper.Callback() {
        override fun onViewPositionChanged(
            changedView: View,
            left: Int,
            top: Int,
            dx: Int,
            dy: Int
        ) {
            super.onViewPositionChanged(changedView, left, top, dx, dy)

            mDragOffset = if (mDragOffset == top.toFloat() / mDragRange) {
                return
            } else {
                top.toFloat() / mDragRange
            }
            if (mDragOffset > 1) {
                mDragOffset = 1f
            }
            if (mDragOffset < 0) {
                mDragOffset = 0f
            }

            callback?.onViewPositionChanged(left, top, mDragOffset)

            val x = 0
            val y = headerView.height + top
            bodyView.layout(x, y, x + bodyView.measuredWidth, y + bodyView.measuredHeight)

            if (isEnableAlpha) {
                bodyView.alpha = 1 - mDragOffset / 2
            } else {
                if (bodyView.alpha != 1f) {
                    bodyView.alpha = 1f
                }
            }

            if (isMinimized) {
                if (isEnableRevertMaxSize) {
                    headerView.pivotX = headerView.width / 2f
                    headerView.pivotY = headerView.height.toFloat()
                    headerView.scaleX = 1 - mDragOffset / 2
                    headerView.scaleY = 1 - mDragOffset / 2
                }
            } else {
                headerView.pivotX = headerView.width / 2f
                headerView.pivotY = headerView.height.toFloat()
                headerView.scaleX = 1 - mDragOffset / 2
                headerView.scaleY = 1 - mDragOffset / 2
            }
            newSizeWHeaderView = (sizeWHeaderViewOriginal * headerView.scaleX).toInt()
            newSizeHHeaderView = (sizeHHeaderViewOriginal * headerView.scaleY).toInt()
            mCenterX = left + sizeWHeaderViewOriginal / 2
            mCenterY = top + newSizeHHeaderView / 2 + sizeHHeaderViewOriginal - newSizeHHeaderView

            when (mDragOffset) {
                0f -> {
                    // top_left, top, top_right
                    when {
                        left <= -headerView.width / 2 -> {
                            changeState(State.TOP_LEFT)
                        }
                        left >= headerView.width / 2 -> {
                            changeState(State.TOP_RIGHT)
                        }
                        else -> {
                            changeState(State.TOP)
                        }
                    }
                }
                1f -> {
                    // bottom_left, bottom, bottom_right
                    when {
                        left <= -headerView.width / 2 -> {
                            changeState(State.BOTTOM_LEFT)
                        }
                        left >= headerView.width / 2 -> {
                            changeState(State.BOTTOM_RIGHT)
                        }
                        else -> {
                            changeState(State.BOTTOM)
                        }
                    }
                    isMinimized = true
                }
                else -> {
                    // mid_left, mid, mid_right
                    when {
                        left <= -headerView.width / 2 -> {
                            changeState(State.MID_LEFT)
                        }
                        left >= headerView.width / 2 -> {
                            changeState(State.MID_RIGHT)
                        }
                        else -> {
                            changeState(State.MID)
                        }
                    }
                }
            }
            if (mCenterY < screenH / 2) {
                if (mCenterX < screenW / 2) {
                    changePart(Part.TOP_LEFT)
                } else {
                    changePart(Part.TOP_RIGHT)
                }
            } else {
                if (mCenterX < screenW / 2) {
                    changePart(Part.BOTTOM_LEFT)
                } else {
                    changePart(Part.BOTTOM_RIGHT)
                }
            }
        }

        override fun tryCaptureView(child: View, pointerId: Int): Boolean {
            return headerView === child
        }

        override fun clampViewPositionVertical(child: View, top: Int, dy: Int): Int {
            val minY: Int = if (isEnableRevertMaxSize) {
                -child.height / 2
            } else {
                -sizeHHeaderViewMin * 3 / 2
            }
            val scaledY = child.scaleY
            val sizeHScaled = (scaledY * child.height).toInt()
            val maxY = height - sizeHScaled * 3 / 2
            return when {
                top <= minY -> {
                    minY
                }
                top > maxY -> {
                    maxY
                }
                else -> {
                    top
                }
            }
        }

        override fun clampViewPositionHorizontal(child: View, left: Int, dx: Int): Int {
            val minX = -child.width / 2
            val maxX = width - child.width / 2
            return when {
                left <= minX -> {
                    minX
                }
                left > maxX -> {
                    maxX
                }
                else -> {
                    left
                }
            }
        }

        override fun onViewReleased(releasedChild: View, xvel: Float, yvel: Float) {
            if (releasedChild === headerView) {
                mViewDragHelper.settleCapturedViewAt(mAutoBackViewX, mAutoBackViewY)
            }
            invalidate()
        }

        override fun onEdgeDragStarted(edgeFlags: Int, pointerId: Int) {}
    }

    private fun changeState(newState: State) {
        if (state != newState) {
            state = newState
            state?.let {
                callback?.onStateChange(it)
            }
        }
    }

    private fun changePart(newPart: Part) {
        if (part != newPart) {
            part = newPart
            part?.let {
                callback?.onPartChange(it)
            }
        }
    }

    override fun computeScroll() {
        if (mViewDragHelper.continueSettling(true)) {
            ViewCompat.postInvalidateOnAnimation(this)
        }
    }

    override fun onInterceptTouchEvent(ev: MotionEvent): Boolean {
        return mViewDragHelper.shouldInterceptTouchEvent(ev)
    }

    @SuppressLint("ClickableViewAccessibility")
    override fun onTouchEvent(event: MotionEvent): Boolean {
        mViewDragHelper.processTouchEvent(event)
        val x = event.x
        val y = event.y
        val isViewUnder = mViewDragHelper.isViewUnder(headerView, x.toInt(), y.toInt())
        when (event.action and MotionEventCompat.ACTION_MASK) {
            MotionEvent.ACTION_UP -> {
                if (state == State.TOP_LEFT || state == State.TOP_RIGHT || state == State.BOTTOM_LEFT || state == State.BOTTOM_RIGHT) {
                    callback?.onOverScroll(state = state, part = part)
                } else {
                    if (part == Part.BOTTOM_LEFT) {
                        minimizeBottomLeft()
                    } else if (part == Part.BOTTOM_RIGHT) {
                        minimizeBottomRight()
                    } else if (part == Part.TOP_LEFT) {
                        if (isEnableRevertMaxSize) {
                            maximize()
                        } else {
                            if (isMinimized) {
                                minimizeTopLeft()
                            }
                        }
                    } else if (part == Part.TOP_RIGHT) {
                        if (isEnableRevertMaxSize) {
                            maximize()
                        } else {
                            if (isMinimized) {
                                minimizeTopRight()
                            }
                        }
                    }
                }
            }
        }
        return isViewUnder
    }

    override fun onLayout(
        changed: Boolean,
        l: Int,
        t: Int,
        r: Int,
        b: Int
    ) {
        super.onLayout(changed, l, t, r, b)
        mDragRange = height - headerView.height
        mAutoBackViewX = headerView.left
        mAutoBackViewY = headerView.top
    }

    enum class State {
        TOP, TOP_LEFT, TOP_RIGHT, BOTTOM, BOTTOM_LEFT, BOTTOM_RIGHT, MID, MID_LEFT, MID_RIGHT
    }

    enum class Part {
        TOP_LEFT, TOP_RIGHT, BOTTOM_LEFT, BOTTOM_RIGHT
    }

    private var state: State? = null
    private var part: Part? = null

    fun maximize() {
        if (isEnableRevertMaxSize) {
            smoothSlideTo(0, 0)
        }
    }

    fun minimizeBottomLeft() {
        val posX = width - sizeWHeaderViewOriginal - sizeWHeaderViewMin / 2
        val posY = height - sizeHHeaderViewOriginal
        smoothSlideTo(posX, posY)
    }

    fun minimizeBottomRight() {
        val posX = width - sizeWHeaderViewOriginal + sizeWHeaderViewMin / 2
        val posY = height - sizeHHeaderViewOriginal
        smoothSlideTo(posX, posY)
    }

    fun minimizeTopRight() {
        if (isEnableRevertMaxSize) {
//            Log.e(TAG, "Error: cannot minimizeTopRight because isEnableRevertMaxSize is true")
            return
        }
        if (!isMinimized) {
//            Log.e(TAG, "Error: cannot minimizeTopRight because isMinimized is false. This function only works if the header view is scrolled BOTTOM")
            return
        }
        val posX = screenW - sizeWHeaderViewMin * 3 / 2
        val posY = -sizeHHeaderViewMin
        smoothSlideTo(posX, posY)
    }

    fun minimizeTopLeft() {
        if (isEnableRevertMaxSize) {
//            Log.e(TAG, "Error: cannot minimizeTopRight because isEnableRevertMaxSize is true")
            return
        }
        if (!isMinimized) {
//            Log.e(TAG, "Error: cannot minimizeTopRight because isMinimized is false. This function only works if the header view is scrolled BOTTOM")
            return
        }
        val posX = -sizeWHeaderViewMin / 2
        val posY = -sizeHHeaderViewMin
        smoothSlideTo(positionX = posX, positionY = posY)
    }

    fun smoothSlideTo(
        positionX: Int,
        positionY: Int
    ) {
        if (mViewDragHelper.smoothSlideViewTo(headerView, positionX, positionY)) {
            ViewCompat.postInvalidateOnAnimation(this)
            postInvalidate()
        }
    }

    fun toggleShowHideHeaderView() {
        if (headerView.visibility == View.VISIBLE) {
            headerView.visibility = View.INVISIBLE
        } else {
            headerView.visibility = View.VISIBLE
        }
    }

    fun toggleShowHideBodyView() {
        if (bodyView.visibility == View.VISIBLE) {
            bodyView.visibility = View.INVISIBLE
        } else {
            bodyView.visibility = View.VISIBLE
        }
    }

    fun onPause() {
        if (!isEnableRevertMaxSize) {
            headerView.visibility = View.INVISIBLE
            bodyView.visibility = View.INVISIBLE
            minimizeBottomRight()
        }
    }

    fun dissappear() {}
    fun appear() {}

    init {
        initView()
    }
}
